### PR TITLE
Harden direct-link DHCP against rogue-server leaks; show app version in the UI

### PR DESCRIPTION
**Pull request by NathanNeurotic**
## What & why

Follow-up to the merged direct-link feature (#92, #93). An adversarial multi-agent review of the whole feature confirmed **two rogue-DHCP-safety findings** (verified 3/3 skeptics, zero refutations). Both share one root cause:

> The pre-answer refusals only inspect **local adapter config** (default gateway, DHCP-origin address). The feature's own setup gives the port a static address with DHCP disabled and no gateway — which **erases exactly those signals** — so `classify_adapter` can never again tell that port apart from a genuine direct link. If the NIC is later moved onto a real LAN (re-cabled, or the PC reboots plugged into a switch) the direct link **re-arms and answers DHCP on the real network**: rogue OFFERs, and broadcast NAKs that kick real clients off valid leases.

Codex's earlier `MAX_DISTINCT_MACS = 1` tightening helped but didn't close it — the first DISCOVER still gets an OFFER, and a single rebooting LAN client still gets NAK'd.

## The fixes

**1. Active foreign-DHCP-server probe — the one guard that observes the *wire*, not local config.**
Before serving, and every 300 s while running, the responder acts as a DHCP *client* on the chosen port: it broadcasts a `DISCOVER` and listens for an `OFFER` from any server other than itself. A PS2 is a client and never answers, so a genuine direct link stays quiet and the responder proceeds; a real network's DHCP server answers, and the responder **refuses (exit 3) before offering anyone an address**.
- Empirically verified on Windows 11 that binding `server_ip:68` with `SO_REUSEADDR` works with the DHCP Client service running, and that the broadcast OFFER is received.
- The responder's own probe MAC (locally-administered) is skipped in `handle_packet`, so the periodic re-probe never trips the single-MAC tripwire or draws a self-offer.

**2. Never broadcast a DHCPNAK.**
A single-lease responder is authoritative for exactly one address, so a request for any other address now gets **silence**, never a NAK. A broadcast NAK on a wire that turned out to be a real network would kick its clients off valid leases (INIT-REBOOT/REBINDING carries no server-id, so it can't be filtered like a foreign SELECTING request). A real PS2 just times out and re-DISCOVERs, and we OFFER it our address.

Defense-in-depth intact: adapter-config refusals, the 1-MAC tripwire, specific-bind interface isolation, and rate limiting all remain.

**3. App version in the UI.**
`v0.4.1` now shows in the window title, the always-visible header strip, and the About tab, so testers can read it off the screen without digging.

## Testing

- **65 unit tests** (`tests/test_directlink.py`): probe packet build/parse, foreign-server detection, self-hearing ignored, wrong-xid ignored, silence on non-matching and INIT-REBOOT requests, own-probe-MAC skip, plus the existing topology/subnet/classification/PowerShell coverage.
- **Live loopback checks** (not committed): the real socket probe (foreign server detected → its id; own server-id ignored → None; nobody answering → None) and the real responder loop (`DISCOVER→OFFER`, `REQUEST→ACK`, foreign `REQUEST→silence`).
- **GUI smoke test**: themed launcher boots; a saved direct link whose adapter is gone is not re-armed (no helper started, checkbox off) and cleans up.

## Compatibility note (Modulo)

The direct link is purely IP/DHCP-layer and is orthogonal to UDPFS/Modulo. A Modulo user ticks **both** boxes — "PS2 is plugged directly into this PC" (for the address) and "using Modulo" (for the UDPFS quirk) — and they stack. The probe uses ports 67/68; UDPFS uses `0xF5F6` + its data port; no interference.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

## Summary by CodeRabbit

- **New Features**
  - Added active detection of competing DHCP servers, with immediate startup probing and periodic re-probing during direct-link operation.
  - Added visible application version information across the window title, header, and About screen.
- **Bug Fixes**
  - Changed behavior to stay silent (no broadcast DHCP NAK) for unsupported/off-target address requests.
  - Improved isolation by ignoring the responder’s own synthetic probe traffic.
- **Tests**
  - Expanded coverage for probe packet formation, foreign-offer detection edge cases, and verification that probe traffic is not answered or tracked.

### CODE REVIEW

File: launcher/directlink.py [hunk 1]
```diff
@@ -477,6 +492,155 @@ def _mac_text(mac):
...
+        server_id = socket.inet_ntoa(sid)
+        if server_id == our_ip:
+            return None  # our own responder answered its own probe
+        return server_id
+    # A reply with no server-id but our xid still means *something* is
+    # answering DHCP out there; name it by the sender's siaddr if present.
+    siaddr = struct.unpack("!I", data[20:24])[0]
+    if siaddr:
+        server_id = _int_to_ip(siaddr)
+        return None if server_id == our_ip else server_id
```
[the remark this answers is not in the document]
Comment (NathanNeurotic) [reply launcher/directlink.py]: Fixed in d74bf07 — removed both same-IP exclusions; any reply to our xid now refuses (the responder provably never answers its own probe: single-threaded + probe-MAC drop, and an ICS twin is literally 192.168.137.1). Test replaced with a detection assertion plus a live loopback check.

_🤖 Addressed by [Claude Code](https://claude.com/claude-code)_

File: launcher/directlink.py [hunk 2]
```diff
@@ -477,6 +492,155 @@ def _mac_text(mac):
...
+    Bounded and side-effect-light: it only ever DISCOVERs, never REQUESTs, so
+    no lease is consumed on whatever network this turns out to be -- exactly
+    what any booting client does.
+    """
+    log = log or (lambda _m: None)
+    rx = socket.socket(socket.AF_INET, socket.SOCK_DGRAM)
+    rx.setsockopt(socket.SOL_SOCKET, socket.SO_REUSEADDR, 1)
+    rx.setsockopt(socket.SOL_SOCKET, socket.SO_BROADCAST, 1)
+    try:
+        rx.bind((server_ip, DHCP_CLIENT_PORT))
```
[the remark this answers is not in the document]
Comment (NathanNeurotic) [reply launcher/directlink.py]: Fixed in d74bf07 — the probe's receive socket now mirrors the responder's mode (wildcard_rx=self.mode=='wildcard'); tx stays bound to the adapter's address so the probe only egresses the chosen port, xid filter screens other interfaces' DHCP chatter. Regression test asserts the mirroring per mode.

_🤖 Addressed by [Claude Code](https://claude.com/claude-code)_

File: launcher/directlink.py [hunk 3]
```diff
@@ -477,6 +495,167 @@
...
+    no lease is consumed on whatever network this turns out to be -- exactly
+    what any booting client does.
+    """
+    log = log or (lambda _m: None)
+    rx_addr = "0.0.0.0" if wildcard_rx else server_ip
+    rx = socket.socket(socket.AF_INET, socket.SOCK_DGRAM)
+    rx.setsockopt(socket.SOL_SOCKET, socket.SO_REUSEADDR, 1)
+    rx.setsockopt(socket.SOL_SOCKET, socket.SO_BROADCAST, 1)
+    try:
+        rx.bind((rx_addr, DHCP_CLIENT_PORT))
```
[the remark this answers is not in the document]
Comment (NathanNeurotic) [reply launcher/directlink.py:553]: Intentional and required, not a leak. This `0.0.0.0` bind only executes on the wildcard-receive fallback (`wildcard_rx`), i.e. the rare hosts where a specifically-bound socket does not receive broadcasts — CodeRabbit explicitly asked us to support that path so the foreign-DHCP probe still hears broadcast OFFERs there. The socket is transient (open for a 2–4s probe, then closed), the transmit side stays bound to the adapter address so the probe only egresses the chosen port, and replies are screened by a random per-probe xid. It also fails safe: the only thing a spoofed reply can do is make us REFUSE the direct link (never serve rogue DHCP). Treating code-scanning alert #9 as an accepted false positive.

_🤖 Addressed by [Claude Code](https://claude.com/claude-code)_
